### PR TITLE
Adds support for fulfillments

### PR DESCRIPTION
**Pull request by Mrkbingham**
# Fulfillments
This merge request seeks to add support for fulfillments and creates models that are reflective of the Square data model.  Orders can have Fulfillments, which themselves have Fulfillments Details of varying types (details on the model update below).  This merge request adds the following:
1. Fulfillment and Fulfillment Details models and migrations
2. Fulfillment and Recipient builders
3. SquareRequest builder calls to support updating of fulfillment details related to orders
4. Tests that cover all the individually added models as well as scenarios for creating and adding orders that contain various Fulfillment Detail types.

## Square API support
This adds support via the `SquareRequestBuilder` to update and modify data on square, and ensure it's synced properly with your application.  Some notable assumptions are baked in here:
1. Orders CANNOT have line-item fulfillment data
2. This also means a single order CANNOT have multiple fulfillments.  This is how the Square UI currently works, so this functionality is only available via the API.  As it's currently an edge case - one that could likely be worked around by creating multiple orders, for now, this will be left out with no plans to add it in.

## Model Updates
This creates a polymorphic relationship between orders via the Fulfillment model.  Adding a fulfillment to an order requires both adding a Fulfillment model and a FulfillmentDetails model - which can be one of three types: `DeliveryDetails`, `PickupDetails` or `ShipmentDetails`.  These relationships are handled through the trait `HasFulfillments`.  

## Recipient Addition
The recipients have been added as a distinct model.  This distinction could be useful for adding future support for additional recipient type, for example [Invoice Recipients](https://github.com/square/square-php-sdk/blob/master/doc/models/invoice-recipient.md).

## Adds Order ID to Payment API request (fulfilling the requirements Square has for orders to show up on the Square Dashboard):
This PR contains a minor fix for adding in the order id to the payments API request, so the payment and order are connected within Square's System.  This update exposed a few bugs that have now been addressed:
1. The Tax calculations in `Util::calculateTotalOrderCost/ByModel` were not taking in to account discounts and inclusive taxes properly.  This now makes sure that Additive taxes are calculated based off the Net Price (taking in to account Discounts and Inclusive Taxes).  Added a private `_calculateNetPrice` method to support this.
2. The Order Builder `buildOrderCopyFromModel` method was not properly creating taxes at the order level.  Added `Constants::DEDUCTIBLE_SCOPE_ORDER`.

### CODE REVIEW

**Comment by Mrkbingham** [comment]
Also when you get a chance @NikolaGavric94, I'm not able to run StyleCI for this repo, so granting access there would be greatly appreciated so I can clean up any syntax issues: 
image

**Comment by NikolaGavric94** [comment]
Hey Mark,

I will take a detailed look at this over the upcoming weekend and will get it sorted by next Tuesday, thank you for your effort and willingness to help :)

**Comment by NikolaGavric94** [comment]
Would you be able to add the wiki page changes with the examples, notes and limitations of using fulfillments with this package in this PR?

File: tests/unit/DeliveryDetailsTest.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        // Create the fulfillment details and associate it with the fulfillment
+        $fulfillment->fulfillmentDetails->save();
+        $fulfillment->fulfillmentDetails()->associate($fulfillment->fulfillmentDetails);
+
+        // Associate order with the fulfillment
+        $fulfillment->order()->associate($this->data->order);
+        $fulfillment->save();
+
+        $this->assertInstanceOf(DeliveryDetails::class, $fulfillment->fresh()->fulfillmentDetails);
```
Comment (NikolaGavric94) [inline tests/unit/DeliveryDetailsTest.php:44]: Can we do some negative testing around these models too? This one in particular needs more overall tests

File: tests/unit/FulfillmentTest.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        $fulfillment->order()->associate($order);
+        $fulfillment->save();
+
+        $fulfillment->order()->associate($order);
+
+        // Save the order
+        $order->save();
+
+        $this->assertInstanceOf(Order::class, $fulfillment->order);
+    }
```
Comment (NikolaGavric94) [inline tests/unit/FulfillmentTest.php:132]: Needs more negative tests

File: tests/unit/PickupDetailsTest.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        // Create the fulfillment details and associate it with the fulfillment
+        $fulfillment->fulfillmentDetails->save();
+        $fulfillment->fulfillmentDetails()->associate($fulfillment->fulfillmentDetails);
+
+        // Associate order with the fulfillment
+        $fulfillment->order()->associate($this->data->order);
+        $fulfillment->save();
+
+        $this->assertInstanceOf(PickupDetails::class, $fulfillment->fresh()->fulfillmentDetails);
+    }
```
Comment (NikolaGavric94) [inline tests/unit/PickupDetailsTest.php:45]: Needs more tests in general with additions of negative tests please

File: tests/unit/ShipmentDetailsTest.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        // Create the fulfillment details and associate it with the fulfillment
+        $fulfillment->fulfillmentDetails->save();
+        $fulfillment->fulfillmentDetails()->associate($fulfillment->fulfillmentDetails);
+
+        // Associate order with the fulfillment
+        $fulfillment->order()->associate($this->data->order);
+        $fulfillment->save();
+
+        $this->assertInstanceOf(ShipmentDetails::class, $fulfillment->fresh()->fulfillmentDetails);
+    }
```
Comment (NikolaGavric94) [inline tests/unit/ShipmentDetailsTest.php:45]: Same as for pickupdetailstest class

**Review by NikolaGavric94 — CHANGES_REQUESTED** [review]

File: src/builders/FulfillmentBuilder.php [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        ) {
+            throw new InvalidSquareOrderException('Fulfillment type does not match details', 500);
+        } elseif (
+            $fulfillment->type == Constants::FULFILLMENT_TYPE_SHIPMENT
+            && ! $fulfillment->fulfillmentDetails instanceof ShipmentDetails
+        ) {
+            throw new InvalidSquareOrderException('Fulfillment type does not match details', 500);
+        }
+
+        $fulfillmentObj = $tempFulfillment;
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/FulfillmentBuilder.php:104]: Yeah, that's a good call.  I was copying the format here from the product builder. Not something I think needs changing in this PR, but worth noting that there may be an unnecessary reassignment over here:
https://github.com/NikolaGavric94/laravel-square/blob/master/src/builders/ProductBuilder.php#L167

[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/FulfillmentBuilder.php]: Updated this to use the tempFulfillment throughout the function instead of reassigning here:
https://github.com/Mrkbingham/laravel-square/commit/c295919974780205a92f7116dcf706b628ee707d

File: src/builders/FulfillmentBuilder.php [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+        } elseif ($type == Constants::FULFILLMENT_TYPE_SHIPMENT) {
+            $fulfillmentDetailsCopy = $this->createShipmentDetailsFromArray($fulfillment, $tempFulfillment);
+        } else {
+            throw new InvalidSquareOrderException('Invalid fulfillment type', 500);
+        }
+
+        // Check for recipient
+        $fulfillmentDetailsCopy->recipient = $this->getRecipientFromFulfillmentArray($fulfillment, $type);
+
+        $fulfillmentObj = $tempFulfillment;
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/FulfillmentBuilder.php]: Updated!  No longer reassigning this - commit for reference here:
https://github.com/Mrkbingham/laravel-square/commit/c295919974780205a92f7116dcf706b628ee707d

File: src/builders/FulfillmentBuilder.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/FulfillmentBuilder.php]: Minor tweak on this one as the original refactor here wasn't correct: https://github.com/Mrkbingham/laravel-square/commit/ddb8dc6b4e5c172dc6165a00efb169cd6806ed2f

File: src/builders/OrderBuilder.php [hunk 7]
```diff
@@ -147,6 +152,41 @@ public function buildOrderFromOrderCopy(Model $order, stdClass $orderCopy): Mode
...
+                // If order doesn't have fulfillment
+                if (! $order->hasFulfillment($fulfillment)) {
+                    // A fulfillment cannot exist without a recipient - make sure it's present
+                    if (!$fulfillment->fulfillmentDetails->recipient) {
+                        throw new MissingPropertyException('Recipient is missing from fulfillment details');
+                    }
+                    // Create the recipient
+                    $recipient = $fulfillment->fulfillmentDetails->recipient;
+                    $recipient->save();
+                    unset($fulfillment->fulfillmentDetails->recipient);
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/OrderBuilder.php:171]: The recipient needs to be unset as the `FulfillmentDetails` model has a many-to-one relationship with the `Recipient` model.  Without unsetting, saving the `FulfillmentDetails` models throws this error:
```
Caused by
PDOException: SQLSTATE[HY000]: General error: 1 table nikolag_shipment_details has no column named recipient
```

I've added a comment to clarify the goal of the `unset`

File: src/builders/OrderBuilder.php [hunk 8]
```diff
@@ -147,6 +152,41 @@ public function buildOrderFromOrderCopy(Model $order, stdClass $orderCopy): Mode
...
+
+                    // Associate the recipient with the fulfillment details
+                    $fulfillment->fulfillmentDetails->recipient()->associate($recipient);
+
+                    // Create the fulfillment details
+                    $fulfillment->fulfillmentDetails->save();
+                    $fulfillment->fulfillmentDetails()->associate($fulfillment->fulfillmentDetails);
+                    unset($fulfillment->fulfillmentDetails);
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/OrderBuilder.php]: This does not need to be unset.  I've removed this:
https://github.com/Mrkbingham/laravel-square/commit/977bbb09fd97521667fe12c8df7a461a4bbac374

File: src/builders/RecipientBuilder.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            'address' => 'required',
+        ];
+
+        // If the recipient has a customer_id, we don't need the individual fields
+        $hasCustomerID = Arr::has($recipientData, 'customer_id') && $recipientData['customer_id'] != null;
+
+        if (!$hasCustomerID) {
+            Validator::make($recipient->toArray(), $individualFieldsRules)->validate();
+        } else {
+            // As long as the customer_id is present, we're good
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/RecipientBuilder.php]: I was basing the required data off of Square's model convention:
https://github.com/square/square-php-sdk/blob/master/doc/models/fulfillment-recipient.md

Their notes say:
The ID of the customer associated with the fulfillment.

> If customer_id is provided, the fulfillment recipient's display_name, email_address, and phone_number are automatically populated from the targeted customer profile.

I've updated the comment here with a reference to that link to add more clarity:
https://github.com/Mrkbingham/laravel-square/commit/3fa664ed21858f4430f1a2aeabb15d13229e0da5

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -38,8 +38,6 @@ jobs:
     environment: sandbox
     steps:
       - uses: actions/checkout@v3
-        with:
-          ref: '3.3.x'
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply .github/workflows/main.yml:42]: I removed this in an effort to run actions on my forked repo.  It looks like there's a branching naming convention I might not have adhered to here, so this does bring up a larger question related to the merge: should I rename my branch to 3.3.1?  I may be misunderstanding how actions work, so if you're aware of any other alternative setups that would help me achieve running tests in my fork I'd be happy to hear of any!  

I'll leave this one open for now in case you've got any suggestions - my intention will be to resolve this issue once all the items below have been taken care of (and I'm able to validate my branch is passing PHPUnit tests successfully) - so long-term I don't intend to keep this change.

File: src/builders/SquareRequestBuilder.php [hunk 11]
```diff
@@ -269,6 +366,117 @@ public function buildAppliedTaxes(Collection $taxes): array
...
+        $fulfillmentDeliveryDetails->setCourierPickupWindowDuration($deliveryDetails->courier_pickup_window_duration);
+        $fulfillmentDeliveryDetails->setCourierProviderName($deliveryDetails->courier_provider_name);
+        $fulfillmentDeliveryDetails->setCourierSupportPhoneNumber($deliveryDetails->courier_support_phone_number);
+
+        // Cancellation/rejection data
+        $fulfillmentDeliveryDetails->setCanceledAt($deliveryDetails->canceled_at?->format(Constants::DATE_FORMAT));
+        $fulfillmentDeliveryDetails->setCancelReason($deliveryDetails->cancel_reason);
+        $fulfillmentDeliveryDetails->setRejectedAt($deliveryDetails->rejected_at?->format(Constants::DATE_FORMAT));
+
+        return $fulfillmentDeliveryDetails;
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/SquareRequestBuilder.php:425]: Are you referring to accessors on the individual properties?
https://laravel.com/docs/11.x/eloquent-mutators#defining-an-accessor

I definitely really like that approach, just want to confirm.

File: src/models/DeliveryDetails.php [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        'cancel_reason',
+        'courier_picked_up_at',
+        'courier_pickup_window_duration',
+        'is_no_contact_delivery',
+        'dropoff_notes',
+        'courier_provider_name',
+        'courier_support_phone_number',
+        'square_delivery_id',
+        'external_delivery_id',
+        'managed_deliver',
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/models/DeliveryDetails.php]: Good call - here's the fields I've removed, lmk what you think:
https://github.com/Mrkbingham/laravel-square/commit/6e573b5846119784cf1ef776558a257a1de42500

File: src/models/PickupDetails.php [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        'placed_at',
+        'accepted_at',
+        'rejected_at',
+        'ready_at',
+        'expired_at',
+        'picked_up_at',
+        'canceled_at',
+        'cancel_reason',
+        'is_curbside_pickup',
+        'curbside_pickup_details',
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/models/PickupDetails.php:45]: Cleaned up the PickupDetails fillable array here - lmk what you think:
https://github.com/Mrkbingham/laravel-square/commit/797811fc81edc7185cf2c680bb86f68dc3715e64

File: src/models/Recipient.php [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public $timestamps = true;
+
+    /**
+     * The attributes that are mass assignable.
+     *
+     * @var array
+     */
+    protected $fillable = [
+        'customer_id',
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/models/Recipient.php]: Removed! https://github.com/Mrkbingham/laravel-square/commit/08276661fac9227baf6474f48763c2dd14abf252

File: src/models/ShipmentDetails.php [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    public $timestamps = true;
+
+    /**
+     * The attributes that are mass assignable.
+     *
+     * @var array
+     */
+    protected $fillable = [
+        'recipient_id',
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/models/ShipmentDetails.php]: Cleaned up here: https://github.com/Mrkbingham/laravel-square/commit/f2901b5907be120167b546565bfe018a35058e14

Lmk what you think

**Comment by Mrkbingham** [comment]
> Would you be able to add the wiki page changes with the examples, notes and limitations of using fulfillments with this package in this PR?

Definitely!  I'm heading out for vacation this week, so will get to this when I return, but will be sure to add a wiki page with examples using the existing ones as a reference point.

File: src/traits/HasFulfillments.php [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * Checks if this model already has a specific fulfillment.
+     *
+     * @param  mixed  $product
+     * @return bool
+     */
+    public function hasFulfillment(mixed $fulfillment): bool
+    {
+        $val = is_array($fulfillment)
+            ? (array_key_exists('id', $fulfillment) ? Product::find($fulfillment['id']) : $fulfillment )
```
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/traits/HasFulfillments.php]: :facepalm: There was definitely a copy/paste error here.  Fixed here!
https://github.com/Mrkbingham/laravel-square/commit/cfe65f19839203950a38e530e91da8a2d098bd45

File: tests/unit/FulfillmentTest.php (hunk 2 above)
Comment (Mrkbingham) [reply tests/unit/FulfillmentTest.php:132]: Added more tests, including some square order tests.

File: tests/unit/PickupDetailsTest.php (hunk 3 above)
Comment (Mrkbingham) [reply tests/unit/PickupDetailsTest.php:45]: Added more tests, along with negative tests.

File: tests/unit/ShipmentDetailsTest.php (hunk 4 above)
Comment (Mrkbingham) [reply tests/unit/ShipmentDetailsTest.php:45]: Added more tests here as well!

File: src/builders/OrderBuilder.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (Mrkbingham) [reply src/builders/OrderBuilder.php:173]: Resolving this one as it's been open for a bit with what I think is a pretty good resolution.

**Comment by Mrkbingham** [comment]
@NikolaGavric94 After a bit of a long hiatus from looking at this, I've resumed work again!  I've added quite a few more tests along with negative testing coverage.  Happy to add anything more that you think is missing.  I still need to add the Wiki pages, but outside of that I think everything should be ready here (other than the one open thread left above re: `mail.yml`)

**Comment by Mrkbingham** [comment]
Okay, I've added the wiki page here:
https://github.com/NikolaGavric94/laravel-square/wiki/Examples-with-Fulfillments-(coming-soon)

**Comment by NikolaGavric94** [comment]
I will check the changes this upcoming weekend and we will see where we are at, thanks for all your hard effort @Mrkbingham

**Comment by NikolaGavric94** [comment]
Would you be able to just attach the picture of passing tests here and I'll merge it. Need to get on fixing actions when PRs are made from forks

**Comment by Mrkbingham** [comment]
FYI @NikolaGavric94 I did make this new MR that contains part of the changes included in here.  I really should have done this sooner, but in order to more easily digest some of the changes, I figured this might be helpful if we ever need to go back and do any debugging:

https://github.com/NikolaGavric94/laravel-square/pull/95

**Comment by Mrkbingham** [comment]
Okay, in addition to branching out a subset of this MR - I've also created a "master" branch in my repo called "nikolag_master" that was just synced from the current state of master of this main repo.  I've created MRs in my own repo against that branch, and have linked here to this successful pipeline run:
https://github.com/Mrkbingham/laravel-square/pull/25/checks

There were a few syntax issues, but those are now resolved too!

**Review by NikolaGavric94 — APPROVED** [review]

**Comment by NikolaGavric94** [comment]
@Mrkbingham Would you be able to resolve these conflicts?

**Comment by Mrkbingham** [comment]
@NikolaGavric94 resolved!

**Comment by Mrkbingham** [comment]
@NikolaGavric94 I'm also noting here that in my repo, I've got the env variables defined in order to get the PHPUnit tests passing, and it looks like those are not defined here:
https://github.com/NikolaGavric94/laravel-square/wiki/Installation-guide

I did a little bit of digging before realizing that the tests here are not likely to pass, however, they're passing over here on the replica PR in my fork:
https://github.com/Mrkbingham/laravel-square/pull/25/checks

**Review by NikolaGavric94 — COMMENTED** [review]
@Mrkbingham Can we revisit `fulfillments` with the new version `40` square once that PR goes in?

**Comment by Mrkbingham** [comment]
@NikolaGavric94 Yeah, since the last round of reviews, I've added a handful of changes - primarily to improve DB performance with tables that have a lot of fulfillments in them, along with a few minor bug-fixes.  I definitely think the Laravel 11 and Square library v 40 make the most sense to move in first before spending too much time on any of the other open PRs.

**Comment by Mrkbingham** [comment]
@NikolaGavric94 oh great call - yeah I went ahead and updated the docs over here.  They did have an errant reference to a constant property that was removed.  I've updated the [Order Fulfillment doc](https://github.com/NikolaGavric94/laravel-square/wiki/Order-Fulfillments), the [Webhooks doc](https://github.com/NikolaGavric94/laravel-square/wiki/Webhooks) and the [Square Facade](https://github.com/NikolaGavric94/laravel-square/wiki/Square-Facade) - as well as making note of all of the new pages on the home page.

There were also some missing methods from the Square facade, so I've added them to this PR.

**Comment by Mrkbingham** [comment]
@NikolaGavric94 While implementing this I ran into a data integrity issue that I've resolved by modifying the relationship between fulfillments and recipients.  In short, recipients have a one-to-one relationship with fulfillments, and not a one-to-many as it had previously been written.  All the test are passing, but in case you're curious to review this change I've limited it to a single commit here:
https://github.com/NikolaGavric94/laravel-square/pull/88/commits/347b742a91904681569c2b5f414bc6f2d9ce5491

**Comment by Mrkbingham** [comment]
@NikolaGavric94 didn't realize I left this one as draft - marking it ready now!
